File: dashboard/src/lib/usage.ts
import { toNumber } from "@/lib/format";

export type UsageRow = {
  id: string;
  created_at: string;
  client_id: string;
  department?: string | null;
  project_id: string;
  model: string;
  prompt_tokens: number | null;
  completion_tokens: number | null;
  cost_usd: number | string | null;
  cost_without_proxy?: number | string | null;
  cost_with_proxy?: number | string | null;
  savings_usd?: number | string | null;
  requested_model?: string | null;
  mode_applied?: string | null;
  routed?: boolean | null;
  stop_loss_triggered?: boolean | null;
  routing_reason?: string | null;
  complexity?: string | null;
};

export function isCacheHit(row: UsageRow) {
  return row.mode_applied === "semantic_cache" || row.model === "semantic-cache";
}

export function summarizeUsage(rows: UsageRow[]) {
  return rows.reduce(
    (acc, row) => {
      acc.requests += 1;
      acc.promptTokens += toNumber(row.prompt_tokens);
      acc.completionTokens += toNumber(row.completion_tokens);
      acc.spent += toNumber(row.cost_with_proxy ?? row.cost_usd);
      acc.withoutProxy += toNumber(row.cost_without_proxy);
      acc.savings += toNumber(row.savings_usd);
      if (isCacheHit(row)) {
        acc.cacheHits += 1;
        acc.cacheSavings += toNumber(row.savings_usd);
        acc.cacheTokens += toNumber(row.prompt_tokens) + toNumber(row.completion_tokens);
      }
      return acc;
    },
    {
      requests: 0,
      promptTokens: 0,
      completionTokens: 0,
      spent: 0,
      withoutProxy: 0,
      savings: 0,
      cacheHits: 0,
      cacheSavings: 0,
      cacheTokens: 0,
    },
  );
}

File: backend/usage.py
from datetime import datetime, timezone
from typing import Dict, Optional, Tuple

from supabase import Client

from backend.smart_balance import PREMIUM_MODEL

MODEL_RATES: Dict[str, Tuple[float, float]] = {
    "openai/gpt-4o": (2.50, 10.00),
    "openai/gpt-4o-mini": (0.15, 0.60),
    "openai/gpt-4.1": (2.00, 8.00),
    "openai/gpt-4.1-mini": (0.40, 1.60),
    "anthropic/claude-3.5-sonnet": (3.00, 15.00),
    "anthropic/claude-3.5-haiku": (0.80, 4.00),
    "deepseek/deepseek-chat": (0.14, 0.28),
}
DEFAULT_RATES = (0.50, 1.50)
AI_REQUESTS_TABLE = "ai_requests"


def estimate_cost(model: str, prompt_tokens: int, completion_tokens: int) -> float:
    prompt_rate, completion_rate = MODEL_RATES.get(model, DEFAULT_RATES)
    return round(
        (prompt_tokens * prompt_rate / 1_000_000)
        + (completion_tokens * completion_rate / 1_000_000),
        8,
    )


def workspace_costs(
    model: str,
    prompt_tokens: int,
    completion_tokens: int,
    raw_cost: Optional[float],
    requested_model: Optional[str] = None,
) -> Dict[str, float]:
    cost_with_proxy = round(float(raw_cost), 8) if raw_cost is not None else estimate_cost(
        model, prompt_tokens, completion_tokens
    )
    baseline = (requested_model or "").strip() or PREMIUM_MODEL
    cost_without_proxy = estimate_cost(baseline, prompt_tokens, completion_tokens)
    savings_usd = round(max(0.0, cost_without_proxy - cost_with_proxy), 8)
    return {
        "cost_with_proxy": cost_with_proxy,
        "cost_without_proxy": cost_without_proxy,
        "savings_usd": savings_usd,
    }


def persist_workspace_usage(
    supabase: Optional[Client],
    *,
    workspace_id: Optional[str],
    requested_model: str,
    model: str,
    prompt_tokens: int,
    completion_tokens: int,
    cost_with_proxy: float,
    cost_without_proxy: float,
    savings_usd: float,
    routed: bool,
    routing_reason: str,
    complexity: str,
    mode_applied: str = "smart_balance",
) -> bool:
    if not supabase:
        print("Supabase is not configured: ai_requests log skipped")
        return False

    row = {
        "client_id": str(workspace_id or "proxy-test"),
        "project_id": "workspace",
        "model": model,
        "prompt_tokens": prompt_tokens,
        "completion_tokens": completion_tokens,
        "cost_usd": cost_with_proxy,
        "created_at": datetime.now(timezone.utc).isoformat(),
        "requested_model": requested_model,
        "mode_applied": mode_applied,
        "routed": routed,
        "stop_loss_triggered": False,
        "routing_reason": routing_reason,
        "cost_without_proxy": cost_without_proxy,
        "cost_with_proxy": cost_with_proxy,
        "savings_usd": savings_usd,
        "complexity": complexity,
    }
    if workspace_id:
        row["workspace_id"] = workspace_id

    try:
        result = supabase.table(AI_REQUESTS_TABLE).insert(row).execute()
        if result.data:
            print(f"Logged usage to {AI_REQUESTS_TABLE}: {result.data[0].get('id')}")
            return True
    except Exception as exc:
        print(f"Failed to insert {AI_REQUESTS_TABLE}: {exc}")

    try:
        fallback = {
            "client_id": str(workspace_id or "proxy-test"),
            "project_id": "workspace",
            "model": model,
            "prompt_tokens": prompt_tokens,
            "completion_tokens": completion_tokens,
            "cost_usd": cost_with_proxy,
            "created_at": datetime.now(timezone.utc).isoformat(),
        }
        result = supabase.table(AI_REQUESTS_TABLE).insert(fallback).execute()
        if result.data:
            print(f"Logged usage (fallback) to {AI_REQUESTS_TABLE}: {result.data[0].get('id')}")
            return True
    except Exception as fallback_exc:
        print(f"Fallback insert failed: {fallback_exc}")
    return False
